src/Raven.Server/ServerWide/SecretProtection.cs

```diff
@@ -543,10 +543,37 @@ private static void AddCertificateChainToTheUserCertificateAuthority(X509Certifi
                         continue;
 
                     var results = userIntermediateStore.Certificates.Find(X509FindType.FindByThumbprint, cert.Thumbprint, false);
-                    if (results.Count > 0)
+                    if (results.Count == 0)
+                        userIntermediateStore.Add(cert);
+                }
+                
+                // We had a problem where we didn't cleanup the user store in Linux (~/.dotnet/corefx/cryptography/x509stores/ca)
+                // and it exploded with thousands of certificates. This caused ssl handshakes to fail on that machine, because it would timeout when
+                // trying to match one of these certs to validate the chain.
+
+                if (loadedCertificate.SubjectName.Name == null)
+                    return;
+                
+                var cnValue = loadedCertificate.SubjectName.Name.StartsWith("CN=", StringComparison.OrdinalIgnoreCase) 
+                    ? loadedCertificate.SubjectName.Name.Substring(3) 
+                    : loadedCertificate.SubjectName.Name;
+                
+                var existingCerts = userIntermediateStore.Certificates.Find(X509FindType.FindBySubjectName, cnValue, false);
+                foreach (var c in existingCerts)
+                {
+                    if (c.NotAfter.ToUniversalTime() > DateTime.Now && c.NotBefore.ToUniversalTime() < DateTime.Now)
                         continue;
+                    
+                    // Remove all expired certs which have the same subject name as our own
+                    var chain = new X509Chain();
+                    chain.Build(c);
 
-                    userIntermediateStore.Add(cert);
+                    foreach (var element in chain.ChainElements)
+                    {
+                        if (element.Certificate.NotAfter.ToUniversalTime() > DateTime.Now && element.Certificate.NotBefore.ToUniversalTime() < DateTime.Now)
```

Comment:
I think we should use `DateTime.UtcNow` ?
Also it would be better to introduce variable before all of the loops and use one value